File: apps/minecraft/spigot/src/main/java/crabcraft/net/crabUtilities/jade/protocol/provider/block/HopperLockProvider.java
package crabcraft.net.crabUtilities.jade.protocol.provider.block;

import net.minecraft.network.RegistryFriendlyByteBuf;
import net.minecraft.network.codec.ByteBufCodecs;
import net.minecraft.network.codec.StreamCodec;
import net.minecraft.resources.Identifier;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import org.jetbrains.annotations.NotNull;
import crabcraft.net.crabUtilities.jade.protocol.JadeProtocol;
import crabcraft.net.crabUtilities.jade.protocol.accessor.BlockAccessor;
import crabcraft.net.crabUtilities.jade.protocol.provider.StreamServerDataProvider;

public enum HopperLockProvider implements StreamServerDataProvider<BlockAccessor, Boolean> {
    INSTANCE;

    private static final Identifier MC_HOPPER_LOCK = JadeProtocol.mc_id("hopper_lock");

    @Override
    public Boolean streamData(@NotNull BlockAccessor accessor) {
        return !accessor.getBlockState().getValue(BlockStateProperties.ENABLED);
    }

    @Override
    public @NotNull StreamCodec<RegistryFriendlyByteBuf, Boolean> streamCodec() {
        return ByteBufCodecs.BOOL.cast();
    }

    @Override
    public Identifier getUid() {
        return MC_HOPPER_LOCK;
    }

    @Override
    public int getDefaultPriority() {
        return BlockNameProvider.INSTANCE.getDefaultPriority() + 10;
    }
}

File: apps/minecraft/spigot/src/main/kotlin/crabcraft/net/crabUtilities/jade/protocol/provider/block/HopperLockProvider.kt
package crabcraft.net.crabUtilities.jade.protocol.provider.block

import crabcraft.net.crabUtilities.jade.protocol.JadeProtocol
import crabcraft.net.crabUtilities.jade.protocol.accessor.BlockAccessor
import crabcraft.net.crabUtilities.jade.protocol.provider.StreamServerDataProvider
import net.minecraft.network.RegistryFriendlyByteBuf
import net.minecraft.network.codec.ByteBufCodecs
import net.minecraft.network.codec.StreamCodec
import net.minecraft.world.level.block.state.properties.BlockStateProperties

enum class HopperLockProvider : StreamServerDataProvider<BlockAccessor, Boolean> {
    INSTANCE;

    override fun streamData(accessor: BlockAccessor): Boolean? {
        return !accessor.getBlockState().getValue(BlockStateProperties.ENABLED)
    }

    override fun streamCodec(): StreamCodec<RegistryFriendlyByteBuf, Boolean> = ByteBufCodecs.BOOL.cast()

    override fun getUid() = UID

    override fun getDefaultPriority(): Int = BlockNameProvider.INSTANCE.getDefaultPriority() + 10

    companion object {
        private val UID = JadeProtocol.mc_id("hopper_lock")
    }
}
